server/service/src/invoice/outbound_return/generate_outbound_return_lines.rs

```diff
@@ -26,62 +29,178 @@ pub fn generate_outbound_return_lines(
+            let filter = StockLineFilter::new()
+                .item_id(EqualFilter::equal_to(item_id))
+                .is_available(true);
+
+            let return_lines = stock_line_repo
+                .query_by_filter(filter, Some(store_id.to_string()))?
+                .iter()
+                .map(stock_line_to_new_return_line)
+                .collect();
+
+            return_lines
         }
-        None => {}
-    }
+        None => vec![],
+    };
 
-    let stock_lines = StockLineRepository::new(&ctx.connection)
-        .query_by_filter(filter, Some(store_id.to_string()))?;
+    // If a return id is included alongside the item, we should include those existing return lines
+    // for that item
+    let existing_return_lines = match (item_id, return_id) {
+        (Some(item_id), Some(return_id)) => {
+            let lines = InvoiceLineRepository::new(&ctx.connection).query_by_filter(
+                InvoiceLineFilter::new()
+                    .invoice_id(EqualFilter::equal_to(&return_id))
+                    .item_id(EqualFilter::equal_to(&item_id)),
+            )?;
 
-    let return_lines: Vec<OutboundReturnLine> = stock_lines
-        .iter()
-        .map(|stock_line| OutboundReturnLine {
-            id: uuid(),
-            stock_line: stock_line.clone(),
+            lines
+                .iter()
+                .map(|line| invoice_line_to_new_return_line(ctx, store_id, line))
+                .collect::<Result<Vec<OutboundReturnLine>, RepositoryError>>()?
+        }
+        _ => vec![],
+    };
 
-            reason_id: None,
-            note: None,
-            number_of_packs: 0,
+    // Filter out any stock lines that are already included in the existing return lines
+    let new_item_lines = from_item_id
+        .into_iter()
+        .filter(|new_line| {
+            !existing_return_lines.iter().any(|existing_line| {
+                new_line.stock_line.stock_line_row.id == existing_line.stock_line.stock_line_row.id
+            })
         })
-        .collect();
+        .collect::<Vec<OutboundReturnLine>>();
 
-    // TODO:
-    // if return_id, query for return lines by return id
+    let mut return_lines = from_stock_line_ids;
+    return_lines.extend(existing_return_lines);
+    return_lines.extend(new_item_lines);
 
     Ok(ListResult {
         count: return_lines.len() as u32,
         rows: return_lines,
     })
 }
 
+fn stock_line_to_new_return_line(stock_line: &StockLine) -> OutboundReturnLine {
+    OutboundReturnLine {
+        id: uuid(),
+        stock_line: stock_line.clone(),
+        reason_id: None,
+        note: None,
+        number_of_packs: 0.0,
+    }
+}
+
+fn invoice_line_to_new_return_line(
+    ctx: &ServiceContext,
+    store_id: &str,
+    line: &InvoiceLine,
+) -> Result<OutboundReturnLine, RepositoryError> {
+    let stock_line_id = line
+        .invoice_line_row
+        .stock_line_id
+        .as_ref()
+        .ok_or(RepositoryError::NotFound)?;
```

Comment:
Using RepositoryError here might be a little bit lazy version... perhaps it would be better for me to create a specific service error here? Or try doing these checks in a validate method? Just noticed that most queries don't have any of that, and this mapped easily into the ListError 🙈